scan-asy-tests-cmake.py

```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/env python3
+
+__doc__ = """
+This file scans tests directory and generates a cmake file containing all *.asy tests,
+excluding wce.
+The resulting file is supposed to be checked in and imported by the main cmake build
+scripts.
+"""
+
+import io
+import os
+import pathlib
+
+REPO_ROOT = pathlib.Path(__file__).parent
+TESTS_DIR = REPO_ROOT / "tests"
+GENERATED_TESTS_LIST_CMAKE_FILE_PATH = (
+    REPO_ROOT / "cmake-scripts/generated/asy-tests-list.cmake"
+)
+
+TESTS_REQURING_CMAKE_FEATURES = {"gsl": "ENABLE_GSL", "gc": "ENABLE_GC"}
+
+TESTS_NOT_PART_OF_CORE_CHECKS = {"gc", "gs"}
+
+TESTS_WITH_ARTIFACTS = {"gc": [".eps"], "output": ["circle.eps", "line.eps"]}
+
+EXCLUDED_TESTS = {"bench"}
+
+GENERATED_CMAKE_FILE_COMMENT_HEADER = f"""
+# This file is automatically generated. Do not modify manually.
+# This file is checked in as part of the repo. It is not meant to be ignored.
+#
+# If more tests are added, run {pathlib.Path(__file__).name} to re-generate
+# the test list.
+"""
+
+
+def generate_tests_list_per_directory(test_dir: os.DirEntry):
+    test_name = test_dir.name
+    if test_name in EXCLUDED_TESTS:
+        return ""
+
+    with os.scandir(test_dir) as scanner_it:
+        tests = [
+            entry.name[:-4]  # removing .asy extension
+            for entry in scanner_it
+            if entry.is_file() and entry.name.endswith(".asy")
+        ]
+
+    if test_name in TESTS_WITH_ARTIFACTS:
+        artifacts_text = f"TEST_ARTIFACTS {' '.join(TESTS_WITH_ARTIFACTS[test_name])}"
+    else:
+        artifacts_text = ""
+
+    test_not_check_str = (
+        "TEST_NOT_PART_OF_CHECK_TEST true"
+        if test_name in TESTS_NOT_PART_OF_CORE_CHECKS
+        else ""
+    )
+
+    with io.StringIO() as text_writer:
+        if test_name in TESTS_REQURING_CMAKE_FEATURES:
+            text_writer.write(f"if ({TESTS_REQURING_CMAKE_FEATURES[test_name]})\n")
+
+        text_writer.write(
+            f"""
```

Comment:
[textwrap.dedent](https://docs.python.org/3/library/textwrap.html#textwrap.dedent) can be used to make multiline strings like this line up better with the surrounding Python code.

Reply:
I clearly did not explain my suggestion well. The point of textwrap.dedent is that you can indent the lines *inside* the multiline string to match the surrounding python code, and dedent will fix it.

More precisely, dedent will then remove the same number of spaces from the beginning of all lines until at least one line has a non-space character as its first character. See https://stackoverflow.com/a/47112169/2318074.

This is just a suggestion -- if you prefer to have the multiline string flush left, that's your prerogative. But personally I find that indenting the lines inside the string makes the python code more readable, since it doesn't disrupt the block indent structure.

Reply:
Ah okay. Does this look okay now? @charlesstaats